polar-language-server/src/lib.rs

```diff
@@ -264,45 +272,74 @@ impl PolarLanguageServer {
             } else {
                 let tracked_docs = self.documents.keys().map(ToString::to_string);
                 let tracked_docs = tracked_docs.collect::<Vec<_>>().join(", ");
                 log(&format!(
                     "untracked doc: {}\n\tTracked: {}\n\tError: {}",
-                    uri, tracked_docs, e
+                    uri, tracked_docs, d
                 ));
                 None
             }
         })
     }
 
-    /// Create a `Diagnostic` from a `PolarError`, filtering out "ignored" errors.
-    fn diagnostic_from_polar_error(
+    /// Create `Diagnostic` from `polar_core::diagnostic::Diagnostic`, filtering out "ignored"
+    /// diagnostics.
+    fn diagnostic_from_polar_diagnostic(
         &self,
-        e: &PolarError,
-    ) -> Option<(TextDocumentItem, Diagnostic)> {
+        d: PolarDiagnostic,
+    ) -> Vec<(TextDocumentItem, Diagnostic)> {
         use polar_core::error::{ErrorKind::Validation, ValidationError::*};
-        match e.kind {
-            // Ignore errors that depend on app data.
-            Validation(UnregisteredClass { .. }) | Validation(SingletonVariable { .. }) => None,
+        use polar_core::warning::ValidationWarning::UnknownSpecializer;
+
+        // Ignore diagnostics that depend on app data.
+        match &d {
+            PolarDiagnostic::Error(e) => match e.kind {
+                Validation(UnregisteredClass { .. }) | Validation(SingletonVariable { .. }) => {
+                    return vec![];
+                }
+                _ => (),
+            },
+            PolarDiagnostic::Warning(w) if matches!(w.kind, UnknownSpecializer { .. }) => {
+                return vec![];
+            }
+            _ => (),
+        }
 
-            _ => self.document_from_polar_error_context(e).map(|doc| {
+        // NOTE(gj): We stringify the error / warning variant instead of the full `PolarError` /
+        // `PolarWarning` because we don't want source context as part of the error message.
+        let (message, severity) = match &d {
+            PolarDiagnostic::Error(e) => (e.kind.to_string(), DiagnosticSeverity::Error),
+            PolarDiagnostic::Warning(w) => (w.kind.to_string(), DiagnosticSeverity::Warning),
+        };
+
+        // If the diagnostic applies to a single doc, use it; otherwise, default to emitting a
+        // duplicate diagnostic for all docs.
+        let docs = self.document_from_polar_diagnostic_context(&d).map_or_else(
+            || self.documents.values().cloned().collect(),
+            |doc| vec![doc],
+        );
+
+        docs.into_iter()
+            .map(|doc| {
                 let diagnostic = Diagnostic {
-                    range: range_from_polar_error_context(e),
-                    severity: Some(DiagnosticSeverity::Error),
+                    range: range_from_polar_error_context(&d),
+                    severity: Some(severity),
                     source: Some("Polar Language Server".to_owned()),
```

Comment:
I wonder if this should just say `Polar`? Is `rust-analyzer`'s error labeled with `rust` or `rust-analyzer`?

Reply:
<img width="212" alt="image" src="https://user-images.githubusercontent.com/17556281/142689099-523c2756-38b6-483f-8aa5-87661e8a2208.png">

Reply:
This is the description of the `Diagnostic.source` field: https://github.com/microsoft/vscode/blob/c15cb13a383dc9ff2dc0828152e374a6b9ecc2b3/src/vs/workbench/contrib/tasks/common/problemMatcher.ts#L754-L758

I chose `"Polar Language Server"` b/c it seemed analogous to their example of `"super lint"`, but I don't have strong feelings either way